README.md
Okay, I give up.
Just read *Postbox Web Interface/maual/index.html* if you want some help.


changelog.txt

GUIDE TO BULLET POINTS (will always be in this order)
  @ announcement
  - removed something
  + added something
  * changed something
  i bug fix
  ~ internal change
  ? w.i.p/planned feature
  / cancelled/postponed
  ! known bug

--    Alpha 2.0    --

  + Added the emoji Kiss Mark through Writing Hand.

  * The gaeFurry variable is now furtard so the security can still work during
    gay pride month, even if less so at Chick-Fil-A.

  * The Markdown guide in the manual is now a link instead of a frame.

  * The vote cooldown has been increased to four minutes.

  ~ Certain methods are properties instead because I discovered
    getters and setters yesterday.

  ~ Postbox is no longer a function. Postpox.post() must be used instead.

  ~ The translation enjine is now compressed.

  ~ The translation engine now runs in strict mode, so hopefully it
    performs better.

  i The manual now includes the correct post existence time.

  ? Going to start working on making my code more D.R.Y.

--    Alpha 1.5    --

  + Added a new original emoji, :uwu:!

  + Added the emoji Grinning Cat Face through Pouting Cat Face.

  + The Markdown guide section on emoji now includes a smiling cat example.

  * Made some small tweaks to the :poo: emoji.

  * The base post existence is now in minutes instead of seconds.

  * The default base post existence has been increased by half a day.

  * The :ghost: emoji is now less ugly.

  * The :goblin: emoji is temporarily going to be a Terraria goblin until I redraw
    it, because the old one was actually a gnome, not a gnoblin.

  i Actually changed the version number... again...

--   Alpha 1.4.1   --

  + Added a bullet point for cancelled or postponed features.

  + Added the emoji Skull through Robot Face. (That includes Pile of Poo!)
    I think my favorite emoji so far is :ogre:...

  * Logical change is now internal change.

  ~ All selection styles with transparency backgrounds now have no background.

  i You can no longer select the space between UI buttons.

--    Alpha 1.4    --

  + Added the emoji Cowboy Hat Face through Angry Face With Horns. You should be excited,
    because that means the next update will be adding Pile of Poo!

  + The Toggle Emoji Size button now actually conveys what it does visually!

  * All UI buttons have been recolored to make them stand out more. Yes, all... three...

  * Big Emoji is now more dynamic.

  * Buttons are now  E X T R A   L A R G E .

  * Buttons can now be easily recolored.

  * Buttons now have a softer look.

  * Changed a few Showdown options.

  * Once again rewrote the hovernotes on various things.

  * The cooldowns for posting and voting have been increased to 20:00 and 6:00 respectively.

  * The Markdown guide now uses a <code> tag when it refers to the <script> tag.

  i The manual no longer erroneously states that titles are optional.

  ? Considering adding the ability to edit posts using a password. Probably won't happen.

  ? Should native emoji be automatically converted to the Postbox ones?

--    Alpha 1.3    --

  + Added a button to toggle Big Emoji.

  + Added the emoji Smirking Face, Grimacing Face, Lying Face, and
    Face With Rolling Eyes because I forgot them before.

  + Added the emoji Face With Medical Mask through Exploding Head. (That's 13 total.)

  + The sorting algorithm is now inside the engine via the new Postbox.sort().

  + The sorting function can now be accessed via Postbox.sort.SORT_FUNCTION.

  * Fixed an indentation mistake on Alpha 0.8's patch notes.

  * Heading selection is now visible.

  * Renamed :sorrow: to :sorrowful:.

  * The half smile emoji is now less like a smirk.

  * The sleeping emoji isn't hiding a DS under the pillow anymore.

  * The sleepy emoji is now even sleepier.

  * The version number is now visible in the window title.

  * Votes now matter 4x as much to the sorting algorithm.

  ~ Showing the version is now the first thing the web UI script does.

--   Alpha  1.2b   --

  * You can now see the text selection on headings.

  i Admin mode is once again off by default.

  i Removed a console.log.

  i You can't select post dates anymore.

--    Alpha 1.2    --

  - You can no longer fall off the edge of the manual because iT's
    A sAfEtY hAzArD.

  + New method: Postbox.getVoteInfluence().

  + New property: <post>.age

  + The Markdown guide now includes a section on HTML character entities.

  * Posts are now sorted using a "smarter" algorithm.

  * The version number is now less distracting and uses a serif font.

  * Version names are now entirely lowercase.

  i Posts are now actually sorted.

  i The version number is now unselectable and has no animation.

  i The vote buttons no longer use the interactive cursor when unavailable.

--   Alpha 1.1.2   --

  + The post creation GUI now has a quicker, smoother AND nicer animation!

  * Changed the hovernote for the close button on the post creation GUI.

  * The animation for opening and closing the manual takes 26% of the time!

  * The corners on the post area are now 100% more rounded!

  * The press animation on the help button is now 100% faster!

  * T H I C C ened the X on the close button on the post creation GUI.

  i The help button no longer appears above the post creation GUI.

  i The version number isn't  S C R O N C H E D  anymore.

  i You can't select the version number anymore.

--   Alpha 1.1.1   --

  - Clicking on the version number no longer opens the changelog because I'm lazy.

  + Added the Sleeping Face emoji (:sleeping:).

  + The manual now opens in a window kind of thing.

  i Fixed the names of the Minecraft fish emoji in Alpha 1.1's changelog.

  ~ The help button is now on top of everything so you can close the manual.

--    Alpha 1.1    --

  @ Humidity is just a measure of how juicy the air is.

  @ Note that the :minecraft/fish: emoji is a tropical fish (clownfish). For cod,
    use :minecraft/cod:, and for salmon, use :minecraft/salmon:.


Postbox Web Interface/script.js
document.querySelector("title").innerText += (version.innerText = Postbox.version);
bigemoji.disabled = true;

var titleSize = Postbox.titleSize, bodySize = Postbox.bodySize;

inputtitle.placeholder = translate("ui.placeholder.title", inputtitle.minLength = titleSize.min, inputtitle.maxLength = titleSize.max);
inputcontent.placeholder = translate("ui.placeholder.content", inputcontent.minLength = bodySize.min, inputcontent.maxLength = bodySize.max);

var noVoteStyles = document.querySelector("#no-vote-styles");
setTimeout(disableNoVoteStyles, Postbox.voteCooldown * 1000);

function helpTransitionEnd() {
	if (this.style.transform !== "none")
		this.src = "about:blank";
}

help.addEventListener("webkittransitionend", helpTransitionEnd);
help.addEventListener("moztransitionend", helpTransitionEnd);
help.addEventListener("mstransitionend", helpTransitionEnd);
help.addEventListener("otransitionend", helpTransitionEnd);
help.addEventListener("transitionend", helpTransitionEnd);

function render(post) {
	// Render a post.
	var e, f;
	(e = posts.insertBefore(document.createElement("FIELDSET"), posts.children[0])).className = "post";
	e.innerHTML = post.html;
	(f = e.appendChild(document.createElement("P"))).setAttribute("id", "post-id");
	f.innerText = post.identifier;
	(e = e.appendChild(document.createElement("LEGEND"))).innerHTML = "<span class=\"date\">" + post.posted + "</span> " + ((post.title || "").replace(/</g, "&lt;").replace(/>/g, "&gt;") || "<i>Untitled</i>") + "<span title=\"" + pm(post.points) + "\" class=\"vote\"><span class=\"vote-up\">▲</span><span id=\"points\">" + post.points + "</span><span class=\"vote-down\">▼</span></span>";
	e.title = "ID: " + post.identifier;
}

function pm(n) {
	// Plus/minus.
	if (n > 0)
		return "+" + n;
	if (n < 0)
		return "-" + n;
	return String(n)
}

Postbox.removeExpiredPosts();
setInterval(function() {
	Postbox.removeExpiredPosts();
	renderPosts();
}, 300000); // Remove old posts every ~5 minutes.

function disableNoVoteStyles() {
	noVoteStyles.disabled = true;
}

function renderPosts() {
	posts.innerHTML = "";

	var p = [];

	var ids = Postbox.allPostIDs();
	for (var i = 0; i < ids.length; i++)
		try {
			p.push(Postbox.read(ids[i]));
		} catch(error) {
		}
	Postbox.sort(p);

	for (var i = 0; i < p.length; i++) {
		render(p[i]);
	}
}

renderPosts();

addEventListener("click", function(event) {
	if (event.detail === 3) {
		if (event.target.tagName.toUpperCase() === "LI" || event.target.tagName.toUpperCase() === "CODE") {
			document.execCommand("Copy");
			console.info("Copied!");
		}
	} else {
		if (event.target.className.replace("down", "up") === "vote-up" && !Postbox.voteCooldown) {
			if (event.target.className === "vote-down") {
				Postbox.downvote(event.target.parentElement.parentElement.parentElement.querySelector("#post-id").innerText);
				event.target.parentElement.querySelector("#points").innerText--;
			} else {
				Postbox.upvote(event.target.parentElement.parentElement.parentElement.querySelector("#post-id").innerText);
				event.target.parentElement.querySelector("#points").innerText++;
			}
			renderPosts();
			noVoteStyles.disabled = false;
			setTimeout(disableNoVoteStyles, Postbox.voteCooldown * 1000);
		}
	}
});

Postbox Web Interface/postbox-old.js
'use strict';

{
	// Various configurations and things, which are only to be modified by people who are supposed to do that, otherwise you're a gay furry.
	let languageStrings = {
		"admin.warning": "\n~~ if you aren't an admin, your r arted you indent your code with spaces instead of tabs. ~~",
		"gui.nope": "Nope",
		"gui.nope.cooldown": "That has been done too recently. Please wait for %s before trying again.",
		"gui.nonexistent": "That %s does not exist!",
		"gui.nope.tooFewCharacters": "Your post is too short. Use least %s characters.",
		"gui.post.contentPlaceholder": "Speak your mind. Format it using Markdown if you want to. You can use up to %s characters, but don't say anything mean because there's no such thing as unposting!",
		"gui.post.titlePlaceholder": "Optionally, write a descriptive title of at most %s characters.",
		"security.linkXXS": "I am an idiot and I tried to make you run the following Javascript:\n\n",
		"week.day.sunday": "Sunday",
		"week.day.monday": "Monday",
		"week.day.tuesday": "Tuesday",
		"week.day.wednesday": "Wednesday",
		"week.day.thursday": "Thursday",
		"week.day.friday": "Friday",
		"week.day.saturday": "Saturday",
		"week.day.today": "Today"
	}; // Translation strings. They use %s like in Minecraft.
	let postBodyLimit = 960; // The character limit for post content.
	let postBodyMin = 10; // The smallest number of characters a post can have.
	let postCooldown = 600; // The post cooldown time in seconds.
	let postExistence = 86400000; // The number of seconds a post normally exists for.
	let postTitleLimit = 50; // The character limit for post titles.
	let voteCooldown = 300; // The post cooldown time in seconds.
	let voteInfluence = 150; // The number of seconds added per upvote.
	let weekdays = [
		"week.day.sunday",
		"week.day.monday",
		"week.day.tuesday",
		"week.day.wednesday",
		"week.day.thursday",
		"week.day.friday",
		"week.day.saturday",
		"week.day.today"
	]; // The names of each week. The eighth one is shown for days that are today.

	let converter = new showdown.Converter();

	function digits(int, digits = 2, apostrophe = true) {
		// actually not from StackOverflow
		// I wrote this one myself. Pretty unusual, right?
		var n = Math.round(int);
		if (n !== n)
			return "??";
		n = String(n);
		return "0".repeat(digits - n.substring(n.length - digits).length) + n.substring(n.length - digits);
	}

	function methodTranslate(string, data = "%s") {
		// Get a language string by name.
		if (string in languageStrings)
			return String(languageStrings[string]).replace(/%s/g, data);
		return String(string);
	}

	// Security Measure #1: Give only this block access to localStorage.
	// It ain't much, but it's honest work.
	let ls = window.localStorage;
	delete window.localStorage;

	let ascii = String.fromCharCode(0) + "	\n\r !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~€ ‚ƒ„…†‡ˆ‰Š‹Œ Ž  ‘’“”•–—˜™š›œ žŸ ¡¢£¤¥¦§¨©ª«¬­®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ";

	function methodCooldown() {
		// Get the amount of time until anyone can post.
		return Math.max(0, postCooldown * 1000 - (Date.now() - methodLast())) || 0;
	}

	function methodCreate(postText, title = "") {
		// Create a post in memory.
		var c = Math.ceil(methodCooldown() / 1000), t = Date.now(), T = "";
		if (String(postText).length < postBodyMin)
			throw methodTranslate("gui.nope") + ": " + methodTranslate("gui.nope.tooFewCharacters", postBodyMin);
		if (c > 0)
			throw methodTranslate("gui.nope") + ": " + methodTranslate("gui.nope.cooldown", digits(Math.floor(c / 60)) + ":" + digits(Math.floor(c % 60)));
		t = t.toString(16);
		t = "0".repeat(16 - t.length) + t;
		ls.setItem("DATA-pt", t = String.fromCharCode("0x" + t.substring(0, 4)) + String.fromCharCode("0x" + t.substring(4, 8)) + String.fromCharCode("0x" + t.substring(8, 12)) + String.fromCharCode("0x" + t.substring(12, 16)));
		var id = String.fromCharCode(Math.round(32 + Math.random() * 95)) +  String.fromCharCode(Math.round(32 + Math.random() * 95)) + String.fromCharCode(Math.round(32 + Math.random() * 95)) + String.fromCharCode(Math.round(32 + Math.random() * 95));
		ls.setItem("POST-" + id, "1" + ascii[0] + t + String(title).substring(0, postTitleLimit) + ascii[0] + String(postText).substring(0, postBodyLimit));
		return id;
	}

	let methodDownvote;

	function methodGetCharacterLimit() {
		// Get the maximum number of characters you can use in the body of a post.
		return postBodyLimit;
	}

	function methodGetTitleCharacterLimit() {
		// Get the maximum number of characters you can use in the title of a post.
		return postTitleLimit;
	}

	function methodHardReset() {
		// ADMIN: Delete everything.
		for (var i = 0; i < ls.length; ls.removeItem(ls.key(i)));
		return "Everything has been deleted. " + methodTranslate("admin.warning");
	}

	function methodLast() {
		// Get the most recent time somebody posted.
		var d = ls.getItem("DATA-pt") || "";
		return 281474976710656 * d.charCodeAt(0) + 4294967296 * d.charCodeAt(1) + 65536 * d.charCodeAt(2) + d.charCodeAt(3);
	}

	function methodLastVote() {
		// Get the most recent time somebody voted on a post.
		var d = ls.getItem("DATA-vt") || "";
		return 281474976710656 * d.charCodeAt(0) + 4294967296 * d.charCodeAt(1) + 65536 * d.charCodeAt(2) + d.charCodeAt(3);
	}

	function methodList() {
		// Get a list of all the existing post IDs.
		var a = [], i = 0, k;
		for (; i <= ls.length; i++)
			if (k = ls.key(i))
				if (k.substring(0, 5) === "POST-")
					a.push(k.substring(5));
		return a;
	}

	function methodRead(postId) {
		// Read a post from memory.
		var key = "POST-" + postId, data = ls.getItem(key), post = {
			"identifier": postId
		}, dateStart = data.indexOf(ascii[0]) + 1, now = new Date();
		post.raw = data;
		post.points = Math.floor(data.substring(0, dateStart - 1)) || 0;
		if (data !== ascii[0]) {
			(post.date = new Date).setTime(281474976710656 * data.charCodeAt(dateStart) + 4294967296 * data.charCodeAt(dateStart + 1) + 65536 * data.charCodeAt(dateStart + 2) + data.charCodeAt(dateStart + 3));
		var y = post.date.getFullYear();
		if (y !== now.getFullYear())
			if (Math.floor(y / 100) === Math.floor(now.getFullYear() / 100))
				y = "/'" + digits(y);
			else
				y = "/" + y;
		else
			y = "";
		post.posted = digits(post.date.getMonth() + 1) + "/" + digits(post.date.getDate()) + y + " @ " + digits(post.date.getHours()) + ":" + digits(post.date.getMinutes());
			var titleEnd = data.indexOf(ascii[0], dateStart + 4);
			post.title = data.substring(dateStart + 4, titleEnd).replace(/\n/g, " ");
			post.html = converter.makeHtml((post.markdown = data.substring(titleEnd + 1)).replace(/\</g, "&lt;")).replace(/\<a href="javascript:/g, "<a href=\"data:text/plain," + encodeURIComponent(methodTranslate("security.linkXXS"))).replace(/\<a /g, "<a target=\"_blank\" ");
		} else
			throw "Nope: The post ID '" + postId + "' doesn't exist!";
		return post;
	}

	function methodRemove(post) {
		// Delete a post from memory.
		if (ls.getItem("POST-" + post) === null)
			throw methodTranslate("nope.nonexistent", "post");
		ls.removeItem("POST-" + post);
		return "The post has been removed. " + methodTranslate("admin.warning");
	}

	function methodRemoveExpired() {
		// Delete all expired posts from memory.
		for (var a = 0, e, i = 0, k, p; i < ls.length; i++)
			if ((k = ls.key(i)).substring(0, 5) === "POST-") {
				p = methodRead(k.replace("POST-", ""));
				e = (p.date.getTime() / 1000 + postExistence + p.points * voteInfluence);
				if (Date.now() / 1000 > e) {
					console.log(k + " has been removed because it is expired.");
					ls.removeItem(k);
					a++;
				} else
					console.log(k + " will continue existing for " + (e - Date.now() / 1000) + " more seconds.");
			}
		return a;
	}

	function methodResetCooldown() {
		// ADMIN: Reset the post cooldown.
		ls.removeItem("DATA-pt");
		ls.removeItem("DATA-vt");
		return "The cooldown has beem reset. " + methodTranslate("admin.warning");
	}

	function methodUpvote(post) {
		// Give a post one point.
		var c, d = ls.getItem("POST-" + post), p, t = Date.now(), points;
		if (d === null)
			throw methodTranslate("gui.nope") + ": " + methodTranslate("gui.nope.nonexistent", "post");
		if (c = methodVoteCooldown())
			throw methodTranslate("gui.nope") + ": " + methodTranslate("gui.nope.cooldown", digits(Math.floor(c / 60)) + ":" + digits(Math.floor(c % 60)));
		var t = Date.now().toString(16);
		t = "0".repeat(16 - t.length) + t;
		ls.setItem("DATA-vt", t = String.fromCharCode("0x" + t.substring(0, 4)) + String.fromCharCode("0x" + t.substring(4, 8)) + String.fromCharCode("0x" + t.substring(8, 12)) + String.fromCharCode("0x" + t.substring(12, 16)));
		return points;
	}

	function methodVoteCooldown() {
		// Get the amount of time until anyone can vote.
		return Math.max(0, voteCooldown * 1000 - (Date.now() - methodLastVote())) || 0;
	}

	// Security Measure #4: Declare the API object as local.
	// Again, someone might try to mess with.
	let pb = window.postbox = function createPostboxPost(content, title) {
		// Create a Postbox post.
		return methodCreate(content, title);
	};
	pb.cooldown = methodCooldown;
	pb.create = methodCreate;
	pb.downvote = methodDownvote;
	pb.getCharacterLimit = methodGetCharacterLimit;
	pb.getTitleCharacterLimit = methodGetTitleCharacterLimit;
	// pb.hardReset = methodHardReset; // Security measure #5: admin methods are commented out
	pb.last = methodLast;
	pb.lastVote = methodLastVote;
	pb.list = methodList;
	pb.read = methodRead;
	// pb.remove = methodRemove;
	pb.removeExpired = methodRemoveExpired;
	// pb.resetCooldown = methodResetCooldown;
	pb.translate = methodTranslate;
	pb.upvote = methodUpvote;
	pb.voteCooldown = methodVoteCooldown;

	addEventListener("load", function() {
		// Security Measure #2: Delete all the scripts.
		// Doesn't really do anything, but can reduce memory usage.
		for (let e; e = document.getElementsByTagName("script")[0]; e.parentElement.removeChild(e));
	});methodResetCooldown();window.localStorage=ls;
}

Postbox Web Interface/postbox.js
'use strict';

/* If you edit this file and you aren't supposed to you're a furry. */

/* # Translation Strings */
translate.register({
	"error.cooldown": "You need to wait %s to do that.",
	"error.nope": "Nope.",
	"error.post.titleTooLong": "Your title is %s characters long, but should be at most %s.",
	"error.post.titleTooShort": "Your title is %s characters long, but should be at least %s.",
	"error.post.tooLong": "Your post is %s characters long, but should be at most %s.",
	"error.post.tooShort": "Your post is %s characters long, but should be at least %s.",
	"error.read.cutOff": "The post %s doesn't contain enough data!",
	"error.read.invalidVersion": "The post %s has an invalid version number.",
	"error.read.nonexistent": "The post %s does not exist!",
	"test": "Hello, world!",
	"test.data": "Hello, %s!",
	"test.multiple": "%s, %s.",
	"ui.placeholder.content": "Speak your mind in %s-%s characters. Don't say anything bad; we will find you! (partial joke) You can also use Markdown to format your post. You can read more on formatting in the manual.",
	"ui.placeholder.title": "Write a descriptive title of %s-%s characters."
});

/* # Main */

{
	/* # Configurations */
	let furtard = false; // Unless you're authorized to do this.

	let postBodyMin = 40; // Smallest number of characters posts can have.
	let postBodyMax = 960; // Largest number of characters posts can have.
	let postCooldown = 1200; // The wait period for posting in seconds.
	let postExistence = 2160; // Base post existence in minutes.
	let postIDLength = 2; // The number of characters to use in a post ID.
	let postTitleMin = 12; // Smallest number of characters titles can have.
	let postTitleMax = 60; // Largest number of characters titles can have.

	let upvoteSelf = true; // Whether posts should automatically be upvoted.

	let voteCooldown = 360; // The wait period for voting in seconds.
	let voteInfluence = 240; // Number of seconds added per vote.

	let reverseSandboxStorage = true; // Hide storage from the outside world.

	/* # Constants */
	const NULL = String.fromCharCode(0);

	/* # Declarations */
	let converter = new showdown.Converter({
		"tables": true,
		"parseImageDimensions": true,
		"omitExtraWLInCodeBlocks": true,
		"emoji": true,
		"backslashEscapesHTMLTags": true
	});
	const ls = localStorage;
	let rn = new Date(); // If you don't know what it stands for Google it.

	/* # Helper Functions */
	const digits = function(num, count = 2) {
		return "'".repeat(String(num).length > count) + "0".repeat(Math.max(0, count - String(num).length)) + String(num).substring(String(num).length - count);
	};

	const plural = function(noun, count = 2) {
		if (Math.abs(count) <= 1)
			return String(noun);
		return String(noun) + "s";
	};

	const formatDate = function(date) {
		// Format an amount of time.
		return digits(date.getDate()) + "/" + digits(date.getMonth() + 1) + "/" + digits(date.getFullYear()) + " @ " + digits(date.getHours()) + ":" + digits(date.getMinutes());
	};

	const formatTime = function(ms) {
		// Format an amount of time.
		var t = Number(ms);
		if (t !== t)
			return "meep";
		if (t < 1000)
			return Math.ceil(t) + plural(" millisecond", t);
		if ((t = Math.ceil(t / 1000)) < 60)
			return t + plural(" second", t);
		if ((t = Math.ceil(t / 60)) < 60)
			return t + plural(" minute", t);
		if ((t = Math.ceil(t / 60)) < 24)
			return t + plural(" hour", t);
		if ((t = Math.ceil(t / 24)) < 7)
			return t + plural(" day", t);
		return Math.ceil(t / 7) + plural(" week", Math.ceil(t / 7));
	};

	const stringFromBinary = function(arr) {
		// Uses UTF-16...
		var str = "", i = 0;
		for (; i < arr.length; i += 2)
			str += String.fromCharCode((Number(arr[i]) || 0) * 256 + (Number(arr[i + 1]) || 0));
		return str;
	};

	const stringToBinary = function(str, arr = []) {
		// Uses UTF-16...
		var c, i = 0, s = String(str);
		for (; i < s.length; i++) {
			c = s.charCodeAt(i);
			arr[i * 2] = Math.floor(c / 256);
			arr[i * 2 + 1] = c % 256;
		}
		return arr;
	};

	/* # Admin Methods */
	const H4X0R_L33T = function() {
		throw translate("error.nope");
	};

	const adminHardReset = function() {
		// Completely clear everything.
		ls.clear();
		return !ls.length; // Returns whether everything was cleared.
	};

	const adminResetCooldowns = function() {
		// Reset the cooldowns.
		ls.removeItem("DATA-pt");
		ls.removeItem("DATA-vt");
		return true; // Wasn't sure what to put here.
	};

	const adminWrite = function(id, info) {
		// Produce data in the latest format.
		var data = [ 0, 0 ], i = 0, txt = String(info.title || "");
		var date = info.date.getTime().toString(2);
		date = "0".repeat(48 - date.length) + date;
		for (; i < date.length / 8; i++)
			data[2 + i] = parseInt(date.substring(i * 8, i * 8 + 8), 2);
		var score = info.points.toString(2);
		score = "0".repeat(32 - score.length) + score;
		for (i = 0; i < score.length / 8; i++)
			data[8 + i] = parseInt(score.substring(i * 8, i * 8 + 8), 2);
		data[12] = txt.length - (txt.length % 256);
		data[13] = txt.length % 256;
		for (i = 0; i < txt.length; i++)
			data[14 + i] = txt.charCodeAt(i);
		for (i = 0; i < info.markdown.length; i++)
			data[14 + txt.length + i] = info.markdown.charCodeAt(i);
		ls.setItem("POST-" + id, stringFromBinary(data));
		return true;
	};

	/* # Methods */
	const methodAllPostIDs = function() {
		// What are all of the existing post IDs?
		var ids = [];
		for (var i = 0, k; i < ls.length; i++)
			if ((k = ls.key(i)).substring(0, 5) === "POST-")
				ids.push(k.replace("POST-", ""));
		return ids;
	};

	const methodBodySize = function() {
		// How long can and should my post be?
		return {
			"min": postBodyMin,
			"max": postBodyMax
		};
	};

	const methodDownvote = function(id) {
		// Decrease a post's score by one point.
		var object = methodRead(id), cooldown = propVoteCooldown();
		if (cooldown)
			throw translate("error.cooldown", formatTime(cooldown * 1000));
		ls.setItem("DATA-vt", Date.now().toString(36));
		object.points--;
		adminWrite(id, object);
		return object.points;
	};

	const methodExists = function(id) {
		// Does this post exist?
		return ls.getItem("POST-" + id) !== null;
	};

	const methodGeneratePostID = function() {
		// Generate a new unique post ID.
		var id = "", i;
		while (!id || ls.getItem("POST-" + id))
			for (i = Number(id = ""); i < postIDLength; i++)
				id += String.fromCharCode(Math.round(Math.random() * 94) + 32);
		return id;
	};

	const methodGetVoteInfluence = function() {
		// How many seconds are added by a vote?
		return voteInfluence;
	};

	const methodPost = function(bodyText, titleText = "") {
		// Create a post on Postbox!
		var content = String(bodyText), title = String(titleText), id = methodGeneratePostID(), cooldown = propPostCooldown();
		if (content.length < postBodyMin)
			throw translate("error.post.tooShort", content.length, postBodyMin);
		if (content.length > postBodyMax)
			throw translate("error.post.tooLong", content.length, postBodyMax);
		if (title.length < postTitleMin)
			throw translate("error.post.titleTooShort", title.length, postTitleMin);
		if (title.length > postTitleMax)
			throw translate("error.post.titleTooLong", title.length, postTitleMax);
		if (cooldown)
			throw translate("error.cooldown", formatTime(cooldown * 1000));
		ls.setItem("DATA-pt", Date.now().toString(36));
		adminWrite(id, {
			"title": titleText,
			"markdown": bodyText,
			"points": Number(!!upvoteSelf),
			"date": rn
		});
		if (upvoteSelf)
			ls.setItem("DATA-vt", Date.now().toString(36));
		return id;
	};

	const methodRead = function(id) {
		// Read a specific post, or throw an error if something goes wrong.
		var raw = ls.getItem("POST-" + id), data = stringToBinary(raw), object = {
			"raw": raw,
			"data": data,
			"markdown": "",
			"html": "",
			"points": 0,
			"identifier": String(id)
		}, len, trg, pos = 2; // Format Version
		if (raw === null)
			throw translate("error.read.nonexistent", id);
		// Format versions are already 16-bit in case I update it 256 times.
		switch (data[0] * 256 + data[1]) {
			case 0:
			if (data.length < 14) // 2b for version, 6b for date, 4b for score, 2b for title length
				throw translate("error.read.cutOff");
			(object.date = new Date()).setTime(data[pos] * 1099511627776 + data[pos + 1] * 4294967296 + data[pos + 2] * 16777216 + data[pos + 3] * 65536 + data[pos + 4] * 256 + data[pos + 5]);
			pos += 6;
			object.points = (data[pos] * 16777216 + data[pos + 1] * 65536 + data[pos + 2] * 256 + data[pos + 3]).toString(2);
			object.points = parseInt(("0".repeat(32 - object.points.length) + object.points).replace(/^1/, "-"), 2);
			pos += 4; // Score
			trg = pos + 2 + (len = data[pos + 1] + data[pos] * 256);
			pos += 2; // Title Length
			if (len)
				if (trg <= data.length)
					for (object.title = ""; pos < trg; object.title += String.fromCharCode(data[pos++])); // Title Text
				else
					throw translate("error.read.cutOff", id);
			for (; pos < data.length; object.markdown += String.fromCharCode(data[pos++])); // Body Text
			break;
			default:
			throw translate("error.read.invalidVersion", id);
			break;
		}
		object.html = converter.makeHtml(object.markdown);
		object.posted = formatDate(object.date);
		object.removalTime = Number(object.date) + postExistence * 60000 + object.points * voteInfluence;
		object.age = performance.now() / 1000 - object.date.getTime() / 1000;
		return object;
	};

	const methodSort = function(list) {
		list.sort(methodSort.SORT_FUNCTION);
		return list;
	};
	methodSort.SORT_FUNCTION = function youCanChangeThis(a, b) {
		try {
			return (b.age - a.age) - ((b.points * voteInfluence * 8) - (a.points * voteInfluence * 8));
		} catch(error) {
			return -Infinity;
		}
	};

	const methodRemoveExpiredPosts = function() {
		// Remove all expired posts, and return the number removed.
		var removed = 0, end = 0, object, i = 0, k;
		for (; i < ls.length; i++)
			if ((k = ls.key(i)).substring(0, 5) === "POST-")
				try {
					object = methodRead(k.replace("POST-", ""));
					end = Number(object.date) + postExistence * 60000 + object.points * voteInfluence;
					if (end < Date.now())
						ls.removeItem(k);
				} catch(error) {
				}
		return removed;
	};

	const methodUpvote = function(id) {
		// Increase a post's score by one point.
		var object = methodRead(id), cooldown = propVoteCooldown();
		if (cooldown)
			throw translate("error.cooldown", formatTime(cooldown * 1000));
		ls.setItem("DATA-vt", Date.now().toString(36));
		object.points++;
		adminWrite(id, object);
		return object.points;
	};

	let probjBodySize = {};

	const propBodySize = function() {
		// How long can and should my post be?
		if (probjBodySize.min !== postBodyMin || probjBodySize.max !== postSizeMax)
			return probjBodySize = {
				"min": postBodyMin,
				"max": postBodyMax
			};
		return probjTitleSize;
	};

	const propLastPosted = function() {
		// When was the last time anyone posted?
		var d = ls.getItem("DATA-pt");
		if (!d)
			return null;
		return new Date(parseInt(d, 36));
	};

	const propLastVoted = function() {
		// When was the last time anyone voted?
		var d = ls.getItem("DATA-vt");
		if (!d)
			return null;
		return new Date(parseInt(d, 36));
	};

	const propPostCooldown = function() {
		// How many seconds until I can post again?
		return (Math.max(0, postCooldown * 1000 - (Date.now() - parseInt(ls.getItem("DATA-pt") || "0", 36))) || 0) / 1000;
	};

	let probjTitleSize = {};

	const propTitleSize = function() {
		// How long can and should my title be?
		if (probjTitleSize.min !== postTitleMin || probjTitleSize.max !== postTitleMax)
			return probjTitleSize = {
				"min": postTitleMin,
				"max": postTitleMax
			};
		return probjTitleSize;
	};

	const propVoteCooldown = function() {
		// How many seconds until I can vote again?
		return (Math.max(0, voteCooldown * 1000 - (Date.now() - parseInt(ls.getItem("DATA-vt") || "0", 36))) || 0) / 1000;
	};

	/* # Initialization */
	let pb = window.Postbox = {
		get bodySize() {
			return propBodySize();
		},
		set bodySize(object) {
			H4X0R_L33T();
		},
		get lastPosted() {
			return propLastPosted();
		},
		set lastPosted(date) {
			H4X0R_L33T();
		},
		get lastVoted() {
			return propLastVoted();
		},
		set lastVoted(date) {
			H4X0R_L33T();
		},
		get postCooldown() {
			return propPostCooldown();
		},
		set postCooldown(seconds) {
			H4X0R_L33T();
		},
		get titleSize() {
			return propTitleSize();
		},
		set titleSize(object) {
			H4X0R_L33T();
		},
		get version() {
			return "alpha 2.0";
		},
		set version(code) {
			H4X0R_L33T();
		},
		get voteCooldown() {
			return propVoteCooldown();
		},
		set voteCooldown(seconds) {
			H4X0R_L33T();
		}
	};
	pb.allPostIDs = methodAllPostIDs;
	pb.downvote = methodDownvote;
	pb.exists = methodExists;
	pb.generatePostID = methodGeneratePostID;
	pb.getVoteInfluence = methodGetVoteInfluence;
	pb.post = methodPost;
	pb.removeExpiredPosts = methodRemoveExpiredPosts;
	pb.read = methodRead;
	pb.sort = methodSort;
	pb.upvote = methodUpvote;

	let admin = pb.admin = {
		"hardReset": adminHardReset,
		"resetCooldowns": adminResetCooldowns,
		"write": adminWrite
	};

	if (!furtard) {
		for (var i in admin)
			switch (typeof admin[i]) {
				case "boolean":
				admin[i] = Boolean(Math.round(Math.random()));
				break;
				case "function":
				admin[i] = H4X0R_L33T;
				break;
				case "number":
				admin[i] = NaN;
				break;
				case "string":
				admin[i] = "";
				break;
				case "symbol":
				case "object":
				admin[i] = null;
				break;
			}
	}
	admin.amIIn = function ifYouCallThisFunctionYouWatchCNN() {
		// But not if you're authorized to do so.
		return !!furtard;
	};

	/* # Finalization */
	if (reverseSandboxStorage)
		delete window.localStorage;
}